sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TTLState.scala

```diff
@@ -19,274 +19,575 @@ package org.apache.spark.sql.execution.streaming
+   * this is running.
+   *
+   * (Why? Some cleanup operations leave the state store in an inconsistent state while
+   * they are doing cleanup. For example, they may choose to get an iterator over all
+   * of the existing values, delete all values, and then re-insert only the non-expired
+   * values from the iterator. If a get is issued after the delete happens but before the
+   * re-insertion completes, the get could return null even when the value does actually
+   * exist.)
    *
    * @return number of values cleaned up.
    */
-  def clearExpiredState(): Long
+  def clearExpiredStateForAllKeys(): Long
+
+  /**
+   * When a user calls clear() on a stateful variable, this method is invoked to
+   * clear all of the state for the current (implicit) grouping key. It is responsible
+   * for deleting from the primary index as well as any secondary index(es).
+   *
+   * If a given state variable has to clean up multiple elementKeys (in MapState, for
+   * example, every key in the map is its own elementKey), then this method should
+   * be invoked for each of those keys.
+   */
+  def clearAllStateForElementKey(elementKey: UnsafeRow): Unit
 }
 
 /**
- * Manages the ttl information for user state keyed with a single key (grouping key).
+ * [[OneToManyTTLState]] is an implementation of [[TTLState]] for stateful variables
+ * that associate a single key with multiple values; every value has its own expiration
+ * timestamp.
+ *
+ * We need an efficient way to find all the values that have expired, but we cannot
+ * issue point-wise deletes to the elements, since they are merged together using the
+ * RocksDB StringAppendOperator for merging. As such, we cannot keep a secondary index
+ * on the key (expirationMs, groupingKey, indexInList), since we have no way to delete a
+ * specific indexInList from the RocksDB value. (In the future, we could write a custom
+ * merge operator that can handle tombstones for deleted indexes, but RocksDB doesn't
+ * support custom merge operators written in Java/Scala.)
+ *
+ * Instead, we manage expiration per grouping key instead. Our secondary index will look
+ * like (expirationMs, groupingKey) -> EMPTY_ROW. This way, we can quickly find all the
+ * grouping keys that contain at least one element that has expired.
+ *
+ * There is some trickiness here, though. Suppose we have an element key `k` that
+ * has a list with one value `v1` that expires at time `t1`. Our primary index looks like
+ * k -> [v1]; our secondary index looks like [(t1, k) -> EMPTY_ROW]. Now, we add another
+ * value to the list, `v2`, that expires at time `t2`. The primary index updates to be
+ * k -> [v1, v2]. However, how do we update our secondary index? We already have an entry
+ * in our secondary index for `k`, but it's prefixed with `t1`, which we don't know at the
+ * time of inserting `v2`.
+ *
+ * So, do we:
+ *    1. Blindly add (t2, k) -> EMPTY_ROW to the secondary index?
```

Comment:
I understand what you're saying, except for:

> but semantic is quite different

Just to clarify, with what you're saying, the semantics might be different, but the _functionality_ is the same? Really, it is less a "time-to-live index" and more a "work queue" of keys that should be serviced, in a particular order. Right?

Reply:
Yes, I'm only concerning about how we reason about needs of TTL index in OneToMany, and explain it. TTL index is technically not needed and we don't use it with the same way we use in OneToOne; it's just to avoid redundant code (since TTL index is hopefully having the same format with reverse index). Without clarification, people would think OneToMany will use the TTL index as the same, while it technically doesn't.

Reply:
Agreed, I'll adjust the comments. In terms of naming, it might be better to call it a "work queue" index. But renaming would be a big, messy refactor. Are you ok to keep it named as "TTL" and clarify in the comments?

Reply:
Yes, maybe just clarify in the class doc that we reuse TTL index in here, but it only tracks the min of expirationMs across elements (with concise reasoning that we will perform full scan anyway).